llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp

```diff
@@ -2067,6 +2067,42 @@ Instruction *InstCombinerImpl::visitIntToPtr(IntToPtrInst &CI) {
   return nullptr;
 }
 
+Value *InstCombinerImpl::foldPtrToIntOfGEP(Type *IntTy, Value *Ptr) {
+  // Look through chain of one-use GEPs.
+  Type *PtrTy = Ptr->getType();
+  SmallVector<GEPOperator *> GEPs;
+  while (true) {
+    auto *GEP = dyn_cast<GEPOperator>(Ptr);
+    if (!GEP || !GEP->hasOneUse())
+      break;
+    GEPs.push_back(GEP);
+    Ptr = GEP->getPointerOperand();
+  }
+
+  // Don't handle case where GEP converts from pointer to vector.
+  if (GEPs.empty() || PtrTy != Ptr->getType())
+    return nullptr;
+
+  // Check whether we know the integer value of the base pointer.
+  Value *Res;
+  Type *IdxTy = DL.getIndexType(PtrTy);
+  if (match(Ptr, m_OneUse(m_IntToPtr(m_Value(Res)))) &&
```

Comment:
Missing test for the multi-use case. Do we need a one-use constraint on inttoptr?

Reply:
This is pre-existing code tested here: https://github.com/llvm/llvm-project/blob/cd7497be6da6a3921dd1148d0590324a625e40e8/llvm/test/Transforms/InstCombine/cast_ptr.ll#L406-L419

I agree that this one-use requirement is unnecessary.

Reply:
It looks like removing this one-use check has some negative interaction with the indexed compare fold.